File: Metal/Front End/Parse/Parser.cs
using Metal.FrontEnd.Parse.Grammar;
using Metal.FrontEnd.Scan;
using System;
using System.Collections.Generic;
using System.ValueType;

namespace Metal.FrontEnd.Parse {
  public class Parser {
    private class ParseError : Exception { }
    private List<Token> tokens;
    private int position = 0;
    private Scanner scanner;
    private Token currentToken;

    public Parser (Scanner scanner) {
      this.tokens = new List<Token>();
      this.scanner = scanner;
    }
    public Parser(List<Token> tokens) {
      this.tokens = tokens;
    }

    public Expression Parse() {
      if (scanner != null) {
        scanner.ScanSafeToken();
        tokens = scanner.Tokens;
      }
      try {
        return ParseExpression();
      } catch(ParseError error) {
        return null;
      }
    }

    private Token Current() {
      return tokens[position];
    }
    private Token Next() {
      if(scanner != null) {
        scanner.ScanSafeToken();
        tokens = scanner.Tokens;
      }
      position++;
      return tokens [position];
    }

    private Expression ParseExpression() {
      return ParseEquality();
    }

    private Expression ParseEquality() {
      Expression expression = ParseComparison();
      while (Match((TokenType.Operator, "!="))) {

      }
      return null;
    }

    private Expression ParseComparison() {
      throw new NotImplementedException();
    }

    private bool Match(params (TokenType, string)[] types) {
      foreach(var type in types) {
        if(Check(type)) {
          Next();
          return true;
        }
      }
      return false;
    }
  }
}

File: metal/frontend/parser/Parser.java
package metal.frontend.parser;

import java.util.List;                                 

import static metal.frontend.scanner.TokenType.*;
import metal.frontend.scanner.Token;
import metal.frontend.parser.grammar.*;

class Parser {                                         
  private final List<Token> tokens;                    
  private int current = 0;                             

  Parser(List<Token> tokens) {                         
    this.tokens = tokens;                              
  } 
  private Expression expression() {
    return equality();       
  }
  private Expression equality() {                         
    Expression expression = comparison();

    while (match('!=', '==')) {        
      Token operator = previous();                  
      Expression right = comparison();                    
      expression = new Expression.Binary(expression, operator, right);
    }                                               

    return expression;                                    
  } 

  private Expression comparison() {                                
    Expression expression = addition();

    while (match('>', '>=', '<', '<=')) {
      Token operator = previous();                           
      Expression right = addition();                               
      expression = new Expression.Binary(expression, operator, right);         
    }                                                        

    return expression;                                             
  } 
  
  private Expression addition() {                         
    Expression expression = multiplication();

    while (match('-', '+')) {                    
      Token operator = previous();                  
      Expression right = multiplication();                
      expression = new Expression.Binary(expression, operator, right);
    }                                               

    return expression;                                    
  }                                                 

  private Expression multiplication() {                   
    Expression expression = unary();                            

    while (match('/', '*')) {                    
      Token operator = previous();                  
      Expression right = unary();                         
      expression = new Expression.Binary(expression, operator, right);
    }                                               

    return expression;                                    
  } 

  private Expression unary() {                     
    if (match('!', '-')) {                
      Token operator = previous();           
      Expression right = unary();                  
      return new Expression.Unary(operator, right);
    }

    return primary();                        
  } 

  private Expression primary() {                                 
    if (match('FALSE')) return new Expression.Literal(false);      
    if (match('TRUE')) return new Expression.Literal(true);        
    if (match('NIL')) return new Expression.Literal(null);

    if (match(NumberLiteral, StringLiteral)) {                           
      return new Expression.Literal(previous().literal);         
    }                                                      

    if (match('(') {                               
      Expression expression = expression();                            
      consume(')', "Expect ')' after expression.");
      return new Expression.Group(expression);                      
    }                                                      
  } 

  private boolean match(TokenType... types) {
    for (TokenType type : types) {           
      if (check(type)) {                     
        advance();                           
        return true;                         
      }                                      
    }

    return false;                            
  } 
  private boolean check(TokenType type) {
    if (isAtEnd()) return false;         
    return peek().type == type;          
  } 
  private Token advance() {   
    if (!isAtEnd()) current++;
    return previous();        
  } 
  private boolean isAtEnd() {      
    return peek().type == EOF;     
  }

  private Token peek() {           
    return tokens.get(current);    
  }                                

  private Token previous() {       
    return tokens.get(current - 1);
  }                                                   
} 